Support torch.sparse_mask on strided input with sparse CSR, CSC, BSR, and BSC mask.

pearu opened the pull request:
While `input.sparse_mask(mask)` can be defined as `input.mul(ones_like(mask))`, implementing this definition leads to a chicken-and-egg problem because the multiplication of dense and sparse_compressed tensors relies on the `sparse_mask` support.

This PR implements `sparse_mask` support for sparse compressed masks using utility functions from sparse compressed tensor conversions support.

Fixes https://github.com/pytorch/pytorch/issues/107373
Fixes https://github.com/pytorch/pytorch/issues/107370

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107903
* #107150
* __->__ #107777
* #107638

pearu committed:
Support torch.sparse_mask on strided input with sparse CSR, CSC, BSR, and BSC mask.

[ghstack-poisoned]

pearu committed:
Update on "Support torch.sparse_mask on strided input with sparse CSR, CSC, BSR, and BSC mask."

Fixes https://github.com/pytorch/pytorch/issues/107373
Fixes https://github.com/pytorch/pytorch/issues/107370




[ghstack-poisoned]

amjames commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -368,20 +370,64 @@ Tensor& fill_sparse_csr_(Tensor& self, const Scalar& value) {
 }
 
-Tensor sparse_mask_sparse_csr(
+Tensor sparse_mask_sparse_compressed(
```
👍

amjames commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -368,20 +370,64 @@ Tensor& fill_sparse_csr_(Tensor& self, const Scalar& value) {
+        });
+  } else if (self.layout() == mask.layout()) {
+    // TODO: keeping this for BC but the method used here may lead to
+    // incorrect indices.
```
This is the bug referenced by #107381 correct?

pearu replied:
yes, exactly. Will need to implement an intersection kernel for sparse compressed indices to fix this.

pearu replied:
closing as OOS for this PR.

pearu committed:
Update on "Support torch.sparse_mask on strided input with sparse CSR, CSC, BSR, and BSC mask."

While `input.sparse_mask(mask)` can be defined as `input.mul(ones_like(mask))`, implementing this definition leads to a chicken-and-egg problem because the multiplication of dense and sparse_compressed tensors relies on the `sparse_mask` support.

This PR implements `sparse_mask` support for sparse compressed masks using utility functions from sparse compressed tensor conversions support.

Fixes https://github.com/pytorch/pytorch/issues/107373
Fixes https://github.com/pytorch/pytorch/issues/107370




[ghstack-poisoned]

pearu committed:
Update on "Support torch.sparse_mask on strided input with sparse CSR, CSC, BSR, and BSC mask."

While `input.sparse_mask(mask)` can be defined as `input.mul(ones_like(mask))`, implementing this definition leads to a chicken-and-egg problem because the multiplication of dense and sparse_compressed tensors relies on the `sparse_mask` support.

This PR implements `sparse_mask` support for sparse compressed masks using utility functions from sparse compressed tensor conversions support.

Fixes https://github.com/pytorch/pytorch/issues/107373
Fixes https://github.com/pytorch/pytorch/issues/107370




[ghstack-poisoned]

pearu committed:
Update on "Support torch.sparse_mask on strided input with sparse CSR, CSC, BSR, and BSC mask."

While `input.sparse_mask(mask)` can be defined as `input.mul(ones_like(mask))`, implementing this definition leads to a chicken-and-egg problem because the multiplication of dense and sparse_compressed tensors relies on the `sparse_mask` support.

This PR implements `sparse_mask` support for sparse compressed masks using utility functions from sparse compressed tensor conversions support.

Fixes https://github.com/pytorch/pytorch/issues/107373
Fixes https://github.com/pytorch/pytorch/issues/107370




[ghstack-poisoned]

pearu committed:
Update on "Support torch.sparse_mask on strided input with sparse CSR, CSC, BSR, and BSC mask."

While `input.sparse_mask(mask)` can be defined as `input.mul(ones_like(mask))`, implementing this definition leads to a chicken-and-egg problem because the multiplication of dense and sparse_compressed tensors relies on the `sparse_mask` support.

This PR implements `sparse_mask` support for sparse compressed masks using utility functions from sparse compressed tensor conversions support.

Fixes https://github.com/pytorch/pytorch/issues/107373
Fixes https://github.com/pytorch/pytorch/issues/107370




[ghstack-poisoned]
